llvm/include/llvm/TargetParser/Triple.h

```diff
@@ -294,7 +294,11 @@ class Triple {
 
     PAuthTest,
 
-    LastEnvironmentType = PAuthTest
+    GNUT64,
+    GNUEABIT64,
+    GNUEABIHFT64,
+
+    LastEnvironmentType = GNUEABIHFT64
```

Comment:
I don't know what the LLVM project's policy is API/ABI stability of marker enum values like this one, but this broke Zig:

```
/home/andy/src/zig/src/zig_llvm.cpp:753:68: error: static assertion failed
  753 | static_assert((Triple::EnvironmentType)ZigLLVM_LastEnvironmentType == Triple::LastEnvironmentType, "");
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/andy/src/zig/src/zig_llvm.cpp:753:68: note: the comparison reduces to ‘(42 == 45)’
```

Reply:
Hmm. That looks a bit odd. The Abi checker didn't catch this. Wonder why - @tstellar

Reply:
Because adding members to an enum generally is not considered an ABI breakage. See e.g.:
https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B#You_can...

Reply:
Ah it was just adding to the end of the enum. That should be fine. It was hard to see the diff on mobile. I was worried it inserted something in the middle of the enum.

Reply:
But in this case, the new members were, in effect, not appended, since `LastEnvironmentType` had its value change. Is the ABI checker deliberately ignoring marker enum members such as this one?

Reply:
Also, ABI aside, quoting [Release Patch Rules](https://llvm.org/docs/HowToReleaseLLVM.html#release-patch-rules):

> Bug fix releases Patches should be limited to bug fixes or very safe and critical performance improvements. Patches must maintain **both API and ABI compatibility** with the previous major release.

Reply:
The change wasn't meant to break API or ABI compatibility, and whether it did is at least controversial. What Zig is doing here is quite uncommon.

Reply:
Yeah this is obviously an oversight and of the ABI checker would have flagged it I would asked for changes.

The question now is what we do about it - reverting the change might not be the best way forward.

I think we need some input from @tstellar @nikic and possibly @AaronBallman here.

Reply:
This landed in 18.1.3 ?

Reply:
> This landed in 19.1.3 ?

Yes

Reply:
I don't *think* a revert at this stage would make much of a difference since the break has already shipped in a release anyway. We'll have to deal with it in Zig regardless.

We just need to be a bit more careful with enum changes like this going forward.

Reply:
I want to understand why the Abi checker didn't flag this. But yes - I will have to read the diffs much more carefully until we know we can trust it.

Maybe it would be enough to have an action that flags any changes to the public include files so that it's not easily missed.

Reply:
@alexrp Why would zig still need to deal with it?  I don't think it's likely 19.1.3 will be used much once 19.1.4 comes out.

Reply:
We don't mandate a particular patch version of LLVM because we try to support building Zig with a distro-provided LLVM. Distro packages aren't necessarily going to be on the latest LLVM patch release.

Reply:
@alexrp At least the distros aren't going to be on 19.1.3 forever.  I don't expect 19.1.3 to be available any more on distros once 19.1.4 ships.

@tru I've seen the ABI checker miss on enums before.  I think we should consider reverting this for 19.1.4.

Reply:
For the record, Gentoo has already backported it to 18.x as well (I was literally waiting for the patch to be merged as a confirmation that this approach is good).

My alternative idea would be to limit the changes to accept `*t64` but return one of the existing enum values, i.e. `-gnut64` would return `::GNU` and so on. We'd lose the ability to customize clang behavior on it, but at least it won't refuse to work right away.

Reply:
This patch doesn't break ABI/API compatibility. The Zig check is overly restrictive and unnecessary. Zig should be fixed instead.

Reply:
The check is intended so that we can find out when LLVM adds additional ABI tags to this enum. For example, it alerted us to this exact change. If the check was deleted as @MaskRay suggests, then we would have missed the fact that we need to add code elsewhere to handle these additional tags.

LLVM should not be modifying the triple in a patch release. This is not a bug fix, it is an enhancement. Just look at the commit message, it begins with the words "Support for ..."

Reply:
Well, that used to be the case anyway. @alexrp pointed out in https://github.com/ziglang/zig/pull/21862 that since we now generate LLVM bitcode rather than using the LLVM IRBuilder API, we can delete these strict checks.

So, this backport is not relevant to the zig project after all.

Reply:
> This patch doesn't break ABI/API compatibility. The Zig check is overly restrictive and unnecessary. Zig should be fixed instead.

It does technically break ABI compatibility, because it changes the value of an enum. This is something we usually try to avoid.

Reply:
Let's continue the discussion and the next steps on discourse, I posted a new post here: https://discourse.llvm.org/t/potential-abi-break-in-19-1-3/82865